src/infer_response.cc

```diff
@@ -167,50 +167,65 @@ InferResponse::Error()
 TRITONSERVER_Error*
 InferResponse::Send(
     TRITONBACKEND_ResponseFactory* response_factory, void* cuda_stream,
-    const uint32_t flags)
+    bool& requires_deferred_callback, const uint32_t flags,
+    std::unique_ptr<SharedMemoryManager>& shm_pool,
+    std::vector<std::pair<std::unique_ptr<PbMemory>, void*>>& output_buffers,
+    const std::set<std::string>& requested_output_names,
+    TRITONBACKEND_Response* response)
 {
-  // [FIXME] Use this code to send responses in non-decoupled mode.
-  TRITONBACKEND_Response* response = nullptr;
-  TRITONSERVER_Error* response_error = nullptr;
-  ScopedDefer response_error_handling([&response, &response_error, flags,
-                                       response_factory] {
-    if (response != nullptr) {
-      LOG_IF_ERROR(
-          TRITONBACKEND_ResponseSend(response, flags, response_error),
-          "failed to send the response.");
-      if (flags == TRITONSERVER_RESPONSE_COMPLETE_FINAL) {
-        std::unique_ptr<
-            TRITONBACKEND_ResponseFactory, backend::ResponseFactoryDeleter>
-        response_factory_ptr(
-            reinterpret_cast<TRITONBACKEND_ResponseFactory*>(response_factory));
-      }
-    }
-  });
+  std::shared_ptr<TRITONSERVER_Error*> response_error =
```

Comment:
This doesn't sound right, when will "if an error is ever set" happen? This is near the end of inference pipeline and why we can't expect error has been set, if any, by this point?

Reply:
Can you clarify on the question? The error can happen any time. This for handling the decoupled responses when the model has returned from the execute function.

Reply:
To my original question, wouldn't this need a custom deleter so that the shared_ptr abstraction actually destructs the TRITONSERVER_Error, instead of just a pointer to the pointer to the object? If there are indeed errors that can happen in the deferred task.

Reply:
Note that if the error does send `TRITONBACKEND_ResponseSend`, must make sure the error in shared pointer is set to nullptr as the error ownership is transferred.

Reply:
@GuanLuo Is it transferred though? As far as I can tell, the error object is not deleted by the Triton core and a copy is created. I think it is the responsibility of the caller to delete the object. https://github.com/triton-inference-server/core/blob/main/src/backend_model.cc#L1029-L1048

Reply:
My bad, that is indeed the case